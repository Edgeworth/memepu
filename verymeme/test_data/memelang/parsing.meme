enum Unit {}
enum TestEnum {
  Test1, Test2, Test3(A),
}
enum Result<R, E> {
  Ok(R), Err(E),
}
fn test<T>(const *T a) Unit {}
struct A { i8 v; bit y : 5; bool z; static fn new() Unit {} }
intf Intf {}
intf IntfT<T> {}
impl Intf for A {}
impl<T> IntfT<T> for A {}
fn main() i8 {
  asm {}
  asm {
    lhu r30, 3
  }
  A decl;
  A st = A {};
  TestEnum test_enum = Test3(st);
  st.v = 0x3;
  i8 e = 0;
  i8 a = 02;
  [40]d = [];
  d[40 - 1] = true && (true || !false) ? ~0 & 1 : 1 ^ 2 | 3;
  if e == st.v || d[23 - 2] != a || a <= e || a > e || e >= st.v || e < d[3] {
    test<i8>(&e);
  } else if true {} else {
    match a {
      1 {}
      2 {}
      3 {print(5);}
    }
    match test_enum {
      Test1 {}
      Test2 {}
      Test3(x) {print(x);}
    }
  }
  const auto b = -(34 + 32) * (55 / 3) - -4;
  a-- 4 % 2;
  a--- b;
  'a';
  '\\';
  '\'';
  '\"';
  "b\"";
  5.5;
  for i8 f = 0; f < 10; f++ {
    return 4;
  }
  // Big comment "\" ''
  return 0;
}
