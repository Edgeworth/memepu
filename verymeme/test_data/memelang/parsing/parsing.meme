enum Unit {}
enum TestEnum {
  Test1, Test2, Test3: A,
}
enum Result<R, E> {
  Ok: R, Err: E,
}
fn test<T>(a: const *T) Unit {}
struct A { v: i8; y: i32; z: bool; static fn new() Unit {} }
intf Intf {}
intf IntfT<T> {}
impl Intf for A {}
impl<T> IntfT<T> for A {}
fn main() i8 {
  asm {}
  asm {
    lhu r30, 3
  }
  var decl: A;
  var st: A = {};
  var test_enum: TestEnum = Test3(st);
  st.v = 0x3;
  var e: i8 = 0;
  var a: i8 = 02;
  var d: [40]i8 = {};
  d[--st.v] = 1;
  d[st.v++] = 1;
  d[40 - 1] = true && (true || !false) ? ~0 & 1 : 1 ^ 2 | 3;
  if (e == st.v || d[23 - 2] != a || a <= e || a > e || e >= st.v || e < d[3]) {
    test<i8>(&e);
  } else if (true) {} else {
    a+b;
  }
  var b: const auto = -(34 + 32) * (55 / 3) - -4;
  a - -4 % 2;
  a--- b;
  'a';
  '\\';
  '\'';
  '\"';
  "b\"";
  5.5;
  a=b=c=d;
  a+b+c+d;
  for (var f: i8 = 0; f < 10; f++) {
    return 4;
  }
  // Big comment "\" ''
  return 0;
}
