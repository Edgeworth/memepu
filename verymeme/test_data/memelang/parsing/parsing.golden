File
 Fn
  FnSig
   Typename(test)
    Typelist(T)
   VarDecl
    VarRef(a)
    Type(T)
     Qualifier(const: 1 ptr: 1, array: 0)
   Type(Unit)
  StmtBlk
 Fn
  FnSig
   Typename(main)
   Type(i8)
  StmtBlk
   Asm()
   Asm(
    lhu r30, 3
  )
   VarDefn
    VarDecl
     VarRef(decl)
     Type(A)
   VarDefn
    VarDecl
     VarRef(st)
     Type(A)
    CompoundLit
   VarDefn
    VarDecl
     VarRef(test_enum)
     Type(TestEnum)
    Op(FN_CALL)
     VarRef(Test3)
     FnCallArgs
      VarRef(st)
   Op(ASSIGNMENT)
    Op(MEMBER_ACCESS)
     VarRef(st)
     VarRef(v)
    IntLit(3)
   VarDefn
    VarDecl
     VarRef(e)
     Type(i8)
    IntLit(0)
   VarDefn
    VarDecl
     VarRef(a)
     Type(i8)
    IntLit(2)
   VarDefn
    VarDecl
     VarRef(d)
     Type(i8)
      Qualifier(const: 0 ptr: 0, array: 40)
    CompoundLit
   Op(ASSIGNMENT)
    Op(ARRAY_ACCESS)
     VarRef(d)
     Op(PREFIX_DEC)
      Op(MEMBER_ACCESS)
       VarRef(st)
       VarRef(v)
    IntLit(1)
   Op(ASSIGNMENT)
    Op(ARRAY_ACCESS)
     VarRef(d)
     Op(POSTFIX_INC)
      Op(MEMBER_ACCESS)
       VarRef(st)
       VarRef(v)
    IntLit(1)
   Op(ASSIGNMENT)
    Op(ARRAY_ACCESS)
     VarRef(d)
     Op(SUB)
      IntLit(40)
      IntLit(1)
    Op(TERNARY)
     Op(LAND)
      BoolLit(true)
      Op(LOR)
       BoolLit(true)
       VarRef(!false)
     Op(TERNARY)
      Op(BAND)
       VarRef(~0)
       IntLit(1)
      Op(BOR)
       Op(BXOR)
        IntLit(1)
        IntLit(2)
       IntLit(3)
   If
    Op(LOR)
     Op(LOR)
      Op(LOR)
       Op(LOR)
        Op(LOR)
         Op(EQ)
          VarRef(e)
          Op(MEMBER_ACCESS)
           VarRef(st)
           VarRef(v)
         Op(NEQ)
          Op(ARRAY_ACCESS)
           VarRef(d)
           Op(SUB)
            IntLit(23)
            IntLit(2)
          VarRef(a)
        Op(LEQ)
         VarRef(a)
         VarRef(e)
       Op(GT)
        VarRef(a)
        VarRef(e)
      Op(GEQ)
       VarRef(e)
       Op(MEMBER_ACCESS)
        VarRef(st)
        VarRef(v)
     Op(LT)
      VarRef(e)
      Op(ARRAY_ACCESS)
       VarRef(d)
       IntLit(3)
    StmtBlk
     Op(GT)
      Op(LT)
       VarRef(test)
       Type(i8)
      Op(UNARY_ADDR)
       VarRef(e)
    If
     BoolLit(true)
     StmtBlk
     StmtBlk
      Op(ADD)
       VarRef(a)
       VarRef(b)
   VarDefn
    VarDecl
     VarRef(b)
     Type(const auto)
    Op(SUB)
     Op(MUL)
      Op(UNARY_NEGATE)
       Op(ADD)
        IntLit(34)
        IntLit(32)
      Op(DIV)
       IntLit(55)
       IntLit(3)
     Op(UNARY_NEGATE)
      IntLit(4)
   Op(SUB)
    VarRef(a)
    Op(MOD)
     Op(UNARY_NEGATE)
      IntLit(4)
     IntLit(2)
   Op(SUB)
    Op(POSTFIX_DEC)
     VarRef(a)
    VarRef(b)
   CharLit(0)
   CharLit(0)
   CharLit(0)
   CharLit(0)
   StrLit(b")
   Op(MEMBER_ACCESS)
    IntLit(5)
    IntLit(5)
   Op(ASSIGNMENT)
    VarRef(a)
    Op(ASSIGNMENT)
     VarRef(b)
     Op(ASSIGNMENT)
      VarRef(c)
      VarRef(d)
   Op(ADD)
    Op(ADD)
     Op(ADD)
      VarRef(a)
      VarRef(b)
     VarRef(c)
    VarRef(d)
   For
    VarDefn
     VarDecl
      VarRef(f)
      Type(i8)
     IntLit(0)
    Op(LT)
     VarRef(f)
     IntLit(10)
    Op(POSTFIX_INC)
     VarRef(f)
    StmtBlk
     Return
      IntLit(4)
   Return
    IntLit(0)
 Enum(0 untyped)
  Typename(Unit)
 Enum(2 untyped)
  Typename(TestEnum)
  VarDecl
   VarRef(Test3)
   Type(A)
 Enum(0 untyped)
  Typename(Result)
   Typelist(R, E)
  VarDecl
   VarRef(Ok)
   Type(R)
  VarDecl
   VarRef(Err)
   Type(E)
 Intf
 Intf
 Struct
  Typename(A)
  VarDecl
   VarRef(v)
   Type(i8)
  VarDecl
   VarRef(y)
   Type(i32)
  VarDecl
   VarRef(z)
   Type(bool)
  Fn
   FnSig
    Typename(new)
    Type(Unit)
   StmtBlk
